packages/web-extension/src/walletManager/walletManagerWorker.ts

```diff
@@ -0,0 +1,129 @@
+import { BehaviorSubject, Observable, filter } from 'rxjs';
+import { Logger } from 'ts-log';
+import { ObservableWallet, storage } from '@cardano-sdk/wallet';
+import { Shutdown, isNotNil } from '@cardano-sdk/util';
+
+import { MessengerDependencies, MinimalRuntime, exposeApi } from '../messaging';
+import {
+  StoresFactory,
+  WalletFactory,
+  WalletManagerActivateProps,
+  WalletManagerApi,
+  WalletManagerProps
+} from './walletManager.types';
+import { consumeKeyAgent } from '../keyAgent';
+import { observableWalletProperties } from '../observableWallet';
+import { walletChannel } from './util';
+
+export interface WalletManagerDependencies {
+  walletFactory: WalletFactory;
+  storesFactory: StoresFactory;
+}
+
+/**
+ * Helper class for background scripts using wallet manager.
+ * Uses wallet and store factories to create wallets.
+ * Keeps track of created stores and reuses them when a wallet is reactivated.
+ */
+export class WalletManagerWorker implements WalletManagerApi {
+  #api$ = new BehaviorSubject<ObservableWallet | null>(null);
+  #hostSubscription: Shutdown;
+  #keyAgentSubscription: Shutdown;
+  #walletStores = new Map<string, storage.WalletStores>();
+
+  #walletFactory: WalletFactory;
+  #storesFactory: StoresFactory;
+  #logger: Logger;
+  #runtime: MinimalRuntime;
+
+  activeWallet$: Observable<ObservableWallet>;
+
+  constructor(
+    { walletName }: WalletManagerProps,
+    { walletFactory, storesFactory, logger, runtime }: MessengerDependencies & WalletManagerDependencies
+  ) {
+    this.activeWallet$ = this.#api$.asObservable().pipe(filter(isNotNil));
+    this.#walletFactory = walletFactory;
+    this.#storesFactory = storesFactory;
+    this.#logger = logger;
+    this.#runtime;
+    this.#hostSubscription = exposeApi(
+      {
+        api$: this.#api$.asObservable(),
+        baseChannel: walletChannel(walletName),
+        properties: observableWalletProperties
+      },
+      { logger, runtime }
+    );
+  }
+
+  async activate(props: WalletManagerActivateProps): Promise<void> {
+    const { observableWalletName } = props;
+    if (await this.#isActive(observableWalletName)) {
+      return;
+    }
+    this.#deactivateWallet();
+
+    // Key agent is created by UI script on a per wallet unique channel
+    const keyAgent = consumeKeyAgent(
+      { walletName: observableWalletName },
+      { logger: this.#logger, runtime: this.#runtime }
+    );
+    this.#keyAgentSubscription = keyAgent;
+    const stores = this.#getStores(observableWalletName);
+
+    // Wallet factory is responsible for creating the wallet and the providers based on cardanoServicesUrl
+    const wallet = await this.#walletFactory.create(props, { keyAgent, stores });
+    this.#api$.next(wallet);
+  }
+
+  async deactivate(): Promise<void> {
+    this.#deactivateWallet();
+  }
+
+  shutdown(): void {
+    this.#deactivateWallet();
+
+    this.#hostSubscription.shutdown();
+    this.#api$.complete();
+    this.#walletStores.clear();
```

Comment:
Stores should only be cleared explicitly
```suggestion
```

Reply:
> Stores should only be cleared explicitly

Once the walletManager is shutdown, it cannot be used anymore, so I do not see the point in keeping the stores, but I may be missing something here.

Reply:
Understood. Updated here: [9f664a4](https://github.com/input-output-hk/cardano-js-sdk/pull/502/commits/9f664a4152ddc7853f8ce6753ba4475f3016187d)